python/pyarrow/dataset.py

```diff
@@ -219,59 +229,48 @@ def source(path_or_paths, filesystem=None, partitioning=None,
         The partitioning scheme specified with the ``partitioning()``
         function. A flavor string can be used as shortcut, and with a list of
         field names a DirectionaryPartitioning will be inferred.
-    format : str
+    format : str, default None
         Currently only "parquet" is supported.
 
     Returns
     -------
     DataSource of DataSourceDiscovery
 
     """
-    filesystem, paths_or_selector = _ensure_fs_and_paths(
-        path_or_paths, filesystem)
-
+    fs, paths_or_selector = _ensure_fs_and_paths(path_or_paths, filesystem)
     partitioning = _ensure_partitioning(partitioning)
-
-    format = format or "parquet"
-    if format == "parquet":
-        format = ParquetFileFormat()
-    elif not isinstance(format, FileFormat):
-        raise ValueError("format '{0}' is not supported".format(format))
+    format = _ensure_format(format or "parquet")
 
     # TODO pass through options
     options = FileSystemFactoryOptions()
-
     if isinstance(partitioning, PartitioningFactory):
         options.partitioning_factory = partitioning
     elif isinstance(partitioning, Partitioning):
         options.partitioning = partitioning
 
-    discovery = FileSystemSourceFactory(
-        filesystem, paths_or_selector, format, options)
-
-    # TODO return Source if a specific schema was passed?
-
-    # need to return SourceFactory since `dataset` might need to
-    # finish the factory with a unified schema
-    return discovery
+    return FileSystemSourceFactory(fs, paths_or_selector, format, options)
 
 
-def _ensure_source(src, filesystem=None, partitioning=None, format=None):
+def _ensure_source(src, **kwargs):
     if _is_path_like(src):
-        src = source(src, filesystem=filesystem,
-                     partitioning=partitioning, format=format)
-    # TODO also accept Source?
-    elif isinstance(src, FileSystemSourceFactory):
-        # when passing a SourceFactory, the arguments cannot be specified
-        if any(kwarg is not None
-               for kwarg in [filesystem, partitioning, format]):
+        return source(src, **kwargs)
+    elif isinstance(src, SourceFactory):
+        if any(v is not None for v in kwargs.values()):
+            # when passing a SourceFactory, the arguments cannot be specified
             raise ValueError(
                 "When passing a Source(Factory), you cannot pass any "
-                "additional arguments")
+                "additional arguments"
+            )
+        return src
+    elif isinstance(src, Source):
+        raise TypeError(
+            "Source objects are currently not supported, only SourceFactory "
+            "instances. Use the ds.source() function to create such objects."
```

Comment:
```suggestion
            "instances. Use the source() function to create such objects."
```

(or in full, but `ds` is not necessarily known to the user)

Reply:
Fixed.